MoltenVK/MoltenVK/GPUObjects/MVKDeviceMemory.h

```diff
@@ -136,41 +128,61 @@ class MVKDeviceMemory : public MVKVulkanAPIDeviceObject {
 					const VkAllocationCallbacks* pAllocator);
 
-    ~MVKDeviceMemory() override;
+	~MVKDeviceMemory() override;
 
 protected:
 	friend class MVKBuffer;
-    friend class MVKImageMemoryBinding;
-    friend class MVKImagePlane;
+	friend class MVKImageMemoryBinding;
+	friend class MVKImagePlane;
 
 	void propagateDebugName() override;
 	VkDeviceSize adjustMemorySize(VkDeviceSize size, VkDeviceSize offset);
-	VkResult addBuffer(MVKBuffer* mvkBuff);
-	void removeBuffer(MVKBuffer* mvkBuff);
-	VkResult addImageMemoryBinding(MVKImageMemoryBinding* mvkImg);
-	void removeImageMemoryBinding(MVKImageMemoryBinding* mvkImg);
-	bool ensureMTLHeap();
-	bool ensureMTLBuffer();
-	bool ensureHostMemory();
-	void freeHostMemory();
-	MVKResource* getDedicatedResource();
-	void initExternalMemory(VkExternalMemoryHandleTypeFlags handleTypes);
-
-	MVKSmallVector<MVKBuffer*, 4> _buffers;
-	MVKSmallVector<MVKImageMemoryBinding*, 4> _imageMemoryBindings;
-	std::mutex _rezLock;
-    VkDeviceSize _allocationSize = 0;
-	MVKMappedMemoryRange _mappedRange;
-	id<MTLBuffer> _mtlBuffer = nil;
+	void checkExternalMemoryRequirements(VkExternalMemoryHandleTypeFlags handleTypes);
+
+	// Backing memory of VkDeviceMemory. This will not be allocated if memory was imported.
+	// Imported memory will directly be backed by MTLBuffer/MTLTexture since there's no way
+	// to create a MTLHeap with existing memory in Metal for now.
 	id<MTLHeap> _mtlHeap = nil;
-	void* _pMemory = nullptr;
-	void* _pHostMemory = nullptr;
+
+	// This MTLBuffer can have 3 usages:
+	// 1. When a heap is allocated, the buffer will extend the whole heap to be able to map and flush memory.
+	// 2. When there's no heap, the buffer will be the backing memory of VkDeviceMemory.
+	// 3. When a texture is imported, the GPU memory will be held by MTLTexture. However, if said texture is
+	// host accessible, we need to provide some memory for the mapping since Metal provides nothing. In this
+	// case, the buffer will hold the host memory that will later be copied to the texture once flushed.
+	id<MTLBuffer> _mtlBuffer = nil;
+
+	// If the user is importing a texture that is not backed by MTLHeap nor MTLBuffer, Metal does not expose
+	// anything to be able to access the texture data such as MTLBuffer::contents. This leads us to having to
+	// use the MTLTexture as the main GPU resource for the memory. If the texture is also host accessible,
+	// a buffer with host visible memory will be allocated as pointed out in point 3 before.
+	id<MTLTexture> _mtlTexture = nil;
+
+	// Mapped memory.
+	void* _map = nullptr;
+	MVKMemoryRange _mapRange = { 0u, 0u };
+
+	// Allocation size.
+	VkDeviceSize _size = 0u;
+	// Metal resource options.
+	MTLResourceOptions _options = 0u;
+
+	// When the allocation is dedicated, it will belong to one specific resource.
+	union {
+		MVKBuffer* _dedicatedBufferOwner = nullptr;
+		MVKImage* _dedicatedImageOwner;
+	};
+	enum class DedicatedResourceType : uint8_t {
+		NONE = 0,
+		BUFFER,
+		IMAGE
```

Comment:
Minor nit. MoltenVK traditional style is to avoid all-caps for names (except in some old macro definitions). So it would be more consistent to call these `None`, `Buffer`, and `Image`.